### PR TITLE
Fix TypeError in sessions -v and add an "unknown" platform class

### LINKED ISSUE #10772: sessions -v crash on nil platform
PR incoming for fix.
When a session is obtained where the platform can't be determined or something bugs out, it becomes `nil`.  However, `sessions -v` can't string concat a `nil` and bombs.

This is a Cisco UC520.
```
msf5 > use auxiliary/scanner/ssh/ssh_login
msf5 auxiliary(scanner/ssh/ssh_login) > set rhosts 2.2.2.2
rhosts => 2.2.2.2
msf5 auxiliary(scanner/ssh/ssh_login) > set username cisco
username => cisco
smsf5 auxiliary(scanner/ssh/ssh_login) > set password cisco
password => cisco
msf5 auxiliary(scanner/ssh/ssh_login) > run

[+] 2.2.2.2:22 - Success: 'cisco:cisco' '  Line has invalid autocommand "id  "'
[*] Command shell session 1 opened (?? -> ??) at 2018-10-08 11:30:14 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/ssh/ssh_login) > sessions -v

[-] Session manipulation failed: no implicit conversion of nil into String ["/root/metasploit-framework/lib/msf/base/serializer/readable_text.rb:772:in `+'", "/root/metasploit-framework/lib/msf/base/serializer/readable_text.rb:772:in `block in dump_sessions_verbose'", "/root/metasploit-framework/lib/msf/core/session_manager.rb:184:in `each'", "/root/metasploit-framework/lib/msf/core/session_manager.rb:184:in `each_sorted'", "/root/metasploit-framework/lib/msf/base/serializer/readable_text.rb:752:in `dump_sessions_verbose'", "/root/metasploit-framework/lib/msf/base/serializer/readable_text.rb:611:in `dump_sessions'", "/root/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:1436:in `cmd_sessions'", "/root/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:501:in `run_command'", "/root/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:453:in `block in run_single'", "/root/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:447:in `each'", "/root/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:447:in `run_single'", "/root/metasploit-framework/lib/rex/ui/text/shell.rb:151:in `run'", "/root/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'", "/root/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'", "./msfconsole:49:in `<main>'"]
```

This is due to the logic that there isn't a default case for `ssh_login`: https://github.com/rapid7/metasploit-framework/blob/master/lib/metasploit/framework/login_scanner/ssh.rb#L152 (yes, i had the last edit on it).

I could argue that this case should have a default of `''` or `'Unknown'`, but i'll at least solve the other issue.

### PR DESCRIPTION
**Pull request by h00die**
Fixes #10772 

To test this you'll need to edit `ssh_login` to return `nil` since I believe this is an edge case.

After patch:
```
msf5 auxiliary(scanner/ssh/ssh_login) > sessions -v

Active sessions
===============

  Session ID: 1
        Name: 
        Type: shell 
        Info: SSH cisco:cisco (2.2.2.2:22)
      Tunnel: ?? -> ?? (2.2.2.2)
         Via: auxiliary/scanner/ssh/ssh_login
   Encrypted: false
        UUID: 
     CheckIn: <none>
  Registered: No

```

While this fixes the `nil` concat issue, @wvu-r7 @bcook-r7 any opinion on what a case select default value should be on ssh_login to prevent a `nil` return?  Empty string or 'Unknown' seem logical to me, (and I can add that on to this PR, just wanted to get more opinions).

### CODE REVIEW

**Comment by wvu-r7** [comment]
Since this is the text output library, I think something like "unknown" is just fine. There is currently no constant for it.

**Comment by wvu-r7** [comment]
Sounds good to me!

**Comment by h00die** [comment]
Case unknown:
```
msf5 auxiliary(scanner/ssh/ssh_login) > run

[+] 2.2.2.2:22 - Success: 'cisco:cisco' '  Line has invalid autocommand "id  "'
[*] Command shell session 1 opened (?? -> ??) at 2018-10-08 13:05:09 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
^[[Amsf5 auxiliary(scanner/ssh/ssh_login) > sessions -v

Active sessions
===============

  Session ID: 1
        Name: 
        Type: shell unknown
        Info: SSH cisco:cisco (2.2.2.2:22)
      Tunnel: ?? -> ?? (2.2.2.2)
         Via: auxiliary/scanner/ssh/ssh_login
   Encrypted: false
        UUID: 
     CheckIn: <none>
  Registered: No
```

also added a new when regex to capture this IOS device as well:
```
msf5 auxiliary(scanner/ssh/ssh_login) > run

[+] 2.2.2.2:22 - Success: 'cisco:cisco' '  Line has invalid autocommand "id  "'
^[[A^[[A[*] Command shell session 1 opened (?? -> ??) at 2018-10-08 13:06:42 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
^[[Amsf5 auxiliary(scanner/ssh/ssh_login) > sessions -v

Active sessions
===============

  Session ID: 1
        Name: 
        Type: shell cisco-ios
        Info: SSH cisco:cisco (2.2.2.2:22)
      Tunnel: ?? -> ?? (2.2.2.2)
         Via: auxiliary/scanner/ssh/ssh_login
   Encrypted: false
        UUID: 
     CheckIn: <none>
  Registered: No
```

File: lib/msf/core/module/platform.rb [hunk 1]
```diff
@@ -576,4 +576,14 @@ class Apple_iOS < Msf::Module::Platform
...
   end
 
+  #
+  # Unknown
+  # This is a special case for when we're completely unsure of the
+  # platform, such as a crash or default case in code.  Only
+  # utilize this as a catch-all.
+  #
+  class Unknown < Msf::Module::Platform
+    Alias = "unknown"
```
Comment (wvu-r7) [inline lib/msf/core/module/platform.rb]: What about `Rank`?

**Review by wvu — CHANGES_REQUESTED** [review]

File: lib/msf/core/module/platform.rb (hunk 1 above)
Comment (h00die) [reply lib/msf/core/module/platform.rb]: https://github.com/rapid7/metasploit-framework/blob/cd9ef01c9a09513d6ec923a732ce770214312920/lib/msf/core/module/platform.rb#L12
inherited, no?

Comment (wvu-r7) [reply lib/msf/core/module/platform.rb]: Yes, but I'd want it to be explicit for this special case.

File: lib/msf/base/serializer/readable_text.rb [hunk 2]
```diff
@@ -768,8 +768,9 @@ def self.dump_sessions_verbose(framework, opts={})
       sess_checkin = "<none>"
       sess_registration = "No"
 
+      # to_s is required incase something bugged earlier and .platform is nil
       if session.respond_to?(:platform)
-        sess_type << " " + session.platform
+        sess_type << " " + session.platform.to_s
```
Comment (wvu-r7) [inline lib/msf/base/serializer/readable_text.rb]: Arguably you shouldn't be appending anything at all if `platform` is `nil`.

**Review by wvu — CHANGES_REQUESTED** [review]

File: lib/msf/core/module/platform.rb (hunk 1 above)
Comment (wvu-r7) [reply lib/msf/core/module/platform.rb]: It's also that I don't trust the default value not to change, and an "unknown" platform adds ambiguity to debugging. :/

File: lib/msf/base/serializer/readable_text.rb (hunk 2 above)
Comment (wvu-r7) [reply lib/msf/base/serializer/readable_text.rb]: Perfect, thanks. I'd just do `if session.respond_to?(:platform) && session.platform`, though. The value is truthy enough to be checked that way without the negative logic.

**Review by wvu — APPROVED** [review]

**Comment by wvu-r7** [comment]
@h00die: See the final changes here: https://github.com/rapid7/metasploit-framework/commit/ccfdfb6a5cdf89f77861cb7117231f36b8db41af. Thank you.

**Comment by wvu-r7** [comment]
# Release Notes

This fixes a `TypeError` crash in the `sessions -v` command for an unknown platform. A new "unknown" platform class has been added for future cases where a platform is truly unknown.
